### DISCUSSION TITLE
MeshCentral certificate issue (locally?) + export logs?

### DISCUSSION THREAD

**Discussion by Skyline643** [Ylianst/MeshCentral#6314] [Issues]
Hello everyone,

I am currently busy to setup a MeshCentral server, with a dynDNS script running for DNS updates.
The translation of the DNS is ok, the hostname is pointing to my WAN address.
But when I follow the Youtube guides and several posts on Github to let the connection with Let'sEncrypt work, I have issues.

My browser keeps telling that the certificate is not valid. But when I tried, I saw the staging message. I already deleted both (production and staging cert) under the letsencrypt certs folder en they renew...

Checks seem to be ok as well (just changed the DNS-name and path of the cert for posting here):

> LE
{
    "configOk": true,
    "leDomains": [
        "my dyn DNS"
    ],
    "challenges": {},
    "production": true,
    "webServer": true,
    "certPath": "some-path/meshcentral-data/letsencrypt-certs",
    "skipChallengeVerification": false,
    "cert": "Present",
    "daysLeft": 89
}
> LECHECK
CertOK
> LEEVENTS
8/8/2024 12:58:07 PM - Getting certs from local store (Production)
8/8/2024 12:58:07 PM - Reading certificate files
8/8/2024 12:58:07 PM - Setting LE cert for default domain.
8/8/2024 12:58:13 PM - Certificate has 89 day(s) left.
8/8/2024 12:58:13 PM - Certificate is ok.
8/8/2024 1:15:25 PM - Certificate has 89 day(s) left.
8/8/2024 1:15:25 PM - Certificate is ok.

Port 80, 443 and default 4433 are open on the server. 4433 and 80 are also open on router and firewall.
When installing (or try to installing...) the agent I got this message:

Resolving somehostname.org (.......org)... 178.X.X.X, 2a02...............................
Connecting to somehostname.org (somehostname.org)|178.X.X.X|:4433... connected.
WARNING: cannot verify somehostname.org's certificate, issued by ‘C=unknown,O=unknown,CN=MeshCentralRoot-643b86’:
  Unable to locally verify the issuer's authority.
HTTP request sent, awaiting response... 404 OK
2024-08-08 15:01:22 ERROR 404: OK.

This is the config.json:

```
{
  "$schema": "https://raw.githubusercontent.com/Ylianst/MeshCentral/master/meshcentral-config-schema.json",
  "__comment1__": "This is a simple configuration file, all values and sections that start with underscore (_) are ignored. Edit a section and remove the _ in front of the name. Refer to the user's guide for details.",
  "__comment2__": "See node_modules/meshcentral/sample-config-advanced.json for a more advanced example.",
  "settings": {
    "mariadb": {"host": "localhost","user": "theadminguy", "password": "verysecretpassword"},
    "cert": "somehostname.org",
    "TrustedCert": false,
    "WANonly": true,
    "_LANonly": true,
    "_sessionKey": "MyReallySecretPassword1",
    "port": 443,
    "aliasPort": 4433,
    "redirPort": 80,
    "_redirAliasPort": 80
  },
  "domains": {
    "": {
      "title": "SomeTitle",
      "title2": "by MeshCentral",
      "_titlePicture": "welcome_1.jpg",
      "welcomePicture": "background.jpg",
      "welcomePictureFullScreen": true,
      "welcomeText": "thecompany",
      "_minify": true,
      "_newAccounts": true,
      "_userNameIsEmail": true,
      "allowedOrigin": true
    }
  },
  "letsencrypt": {
    "__comment__": "Requires NodeJS 8.x or better, Go to https://letsdebug.net/ first before trying Let's Encrypt.",
    "email": "mail@hostname",
    "names": "somehostname.org",
    "skipChallengeVerification": false,
    "rsaKeySize": 3072,
    "production": true,
    "_lib": "acme-client"
  }
}
```

So I hope someone can help me to find what's the issue. Node version is V16.20.2 running on an Ubuntu 22.04.4 server on a Proxmox host. MeshCentral version is the latest one with a separate mariadb database because the hardware wasn't supporting mongodb.

**Comment by si458** [comment]
try removing `"TrustedCert": false` and restarting meshcentral

**Reply by si458** [reply]
also make sure `meshcentral-data/letsencrypt-certs` has certificates in them `production.key` and `production.crt`

**Reply by si458** [reply]
also try running meshcentral with cert debug and watch the output 
`node node_modules/meshcentral --debug cert`

**Reply by Skyline643** [reply]
Thanks for your fast reply. Removing the TrustedCert entry did nothing.
The production.key and production.crt files are present, as well as the staging files. I did a test to rename them to .old and they were regenerated (first the staging, than changed the production value in the config to get the production certs too).



When I run the letsdebug I get the following result:

[AAAANotWorking](https://letsdebug.net/somedomainname.org/2162388#AAAANotWorking-Error)
Error
somedomainname.org has an AAAA (IPv6) record (2a02:ipv6:shiz:zle) but a test request to this address over port 80 did not succeed. Your web server must have at least one working IPv4 or IPv6 address. You should either ensure that validation requests to this domain succeed over IPv6, or remove its AAAA record.
A timeout was experienced while communicating with somedomainname.org/2a02:ipv6thing: Get "http://somedomainname.org/.well-known/acme-challenge/letsdebug-test": context deadline exceeded

Trace:
@0ms: Making a request to http://somedomainname.org/.well-known/acme-challenge/letsdebug-test (using initial IP 2a02:ipv6:thi:ng)
@0ms: Dialing 2a02:ipv6:stu:ff
@10000ms: Experienced error: context deadline exceeded



And this is the result from running meshcentral with cert debug:

MeshCentral HTTP redirection server running on port 80.
CERT: LE: Getting certs from local store (Production)
CERT: LE: Reading certificate files
CERT: LE: Setting LE cert for default domain.
MeshCentral v1.1.27, WAN mode.
MeshCentral Intel(R) AMT server running on somehostname.org:4433.
MeshCentral HTTPS server running on somehostname.org:443, alias port 4433.
CERT: LE: Certificate has 89 day(s) left.
CERT: LE: Certificate is ok.

The ipv6 address is listening on port 80, so I find it strange that dynu is having issues with the AAAA record?

**Reply by si458** [reply]
so is the machine an IPv6 only? or is it 4+6 ?

**Reply by Skyline643** [reply]
On the MeshCentral virtual machine, only ipv4 is configured in netplan. Rules on the ufw firewall are for ipv4 and ipv6 (ipv6 is enabled). But with the ip -6 addr show command I receive a valid ipv6 address. On dynu, the AAAA record is also pointing to the ipv6 address of the virtual machine.

A simple telnet to this ipv6 address on port 80 seems to work. MeshCentral is the only webservice on that machine.
The DNS lookup tool at Dynu is also pointing to that ipv6 address of the virtual machine if I enter the DNS Host.

Very strange. I don't know what might be the problem.

**Reply by si458** [reply]
So u need to check ur firewall on the vm, or completely disable it for the moment whole u get everything working

**Reply by Skyline643** [reply]
I disabled the firewall completely, but still get the same error for the AAAA record.

With a netstat -lntup command I see the server is listening on the necessary ports:

tcp6       0      0 :::80                   :::*                    LISTEN      16010/node
tcp6       0      0 :::443                  :::*                    LISTEN      16010/node
tcp6       0      0 :::4433                 :::*                    LISTEN      16010/node

16010 is the process ID, I think 'node' is the nodejs tool I've installed to let MeshCentral work in the background?

**Reply by si458** [reply]
from the looks of that output, node/meshcentral is ONLY listening on ipv6 and no ipv4?
so now im completely confused WTF is going on?
lets try starting a fresh!
1. backup the letsencrypt folder in meshcentral-data then delete it!
2. make sure firewall stays off completely, this just elimates any issues for now
3. check your IP addresses externally both ipv4 and ipv6
`curl -4 https://api.ipify.org/`
`curl -6 https://api6.ipify.org/`
4. make sure the DNS records for A and AAAA MATCH the ips in step 3
5. remove the letsencrypt section from your config.json (or just comment it out `"_letsencrypt"`
6. restart meshcentral and check the output when it starts up to make sure it says its using 80/443 correctly
7. check its actually listening on 80/443 for ipv4/ipv6 `lsof -Pni`
8. try visiting `https://somedomainname.org` in a private browser and verify its SSL as the meshcentralroot certificate and the commonname matches ur dns name

let us know when u done this

**Reply by Skyline643** [reply]
Ok, thanks for the steps. I followed the steps and everything seems to be ok till step 7. The results of the curl -4 matches the WAN IP of the router, curl -6 matches the IPv6 address of the virtual machine.

When I run step 6, the result seems to be ok. But at step 7 I don't get the right response:

**Seems ok:**
root@XXX:/home/admin# sudo systemctl stop meshcentral.service
root@XXX:/home/admin# node /home/benoit-admin/node_modules/meshcentral --debug cert
MeshCentral HTTP redirection server running on port 80.
MeshCentral v1.1.27, WAN mode.
MeshCentral Intel(R) AMT server running on mydomainname.org:4433.
MeshCentral HTTPS server running on mydomainname:443, alias port 4433.

**Here I see no port 80, 443 or 4433 mentioned?**
root@XXX:/home/admin# lsof -Pni
COMMAND     PID            USER   FD   TYPE DEVICE SIZE/OFF NODE NAME
systemd       1            root  136u  IPv6  21909      0t0  TCP *:9090 (LISTEN)
systemd-n   722 systemd-network   20u  IPv6  21885      0t0  UDP [fe80::78 and so on]:546
systemd-r   724 systemd-resolve   13u  IPv4  20926      0t0  UDP 127.0.0.53:53
systemd-r   724 systemd-resolve   14u  IPv4  20927      0t0  TCP 127.0.0.53:53 (LISTEN)
mariadbd    857           mysql   25u  IPv4  22259      0t0  TCP 127.0.0.1:3306 (LISTEN)
sshd        931            root    3u  IPv4  22736      0t0  TCP *:22 (LISTEN)
sshd      16362            root    4u  IPv4  54836      0t0  TCP 192.X.X.X:22->192.X.X.X:50884 (ESTABLISHED)
sshd      16472    admin    4u  IPv4  54836      0t0  TCP 192.X.X.X:22->192.X.X.X:50884 (ESTABLISHED)
sshd      16477            root    4u  IPv4  54917      0t0  TCP 192.X.X.X:22->192.X.X.X:50892 (ESTABLISHED)
sshd      16527    admin    4u  IPv4  54917      0t0  TCP 192.X.X.X:22->192.X.X.X:50892 (ESTABLISHED)
sshd      16624            root    4u  IPv4  55013      0t0  TCP 192.X.X.X:22->192.X.X.X:51558 (ESTABLISHED)
sshd      16684    admin    4u  IPv4  55013      0t0  TCP 192.X.X.X:22->192.X.X.X:51558 (ESTABLISHED)

**Reply by Skyline643** [reply]
And this is the certificate I see in my browser (but with the message that the certificate is invalid):

![image](https://github.com/user-attachments/assets/fffd4e3e-8062-4343-86e6-d386f584042f)

**Reply by si458** [reply]
OK don't worry about step 7 as that was just to prove it was running in the correct port, but it must be otherwise u wouldn't of been able to get a invalid certificate and the web ui!
So now it's showing an invalid certificate, put the letsencrypt section back in, stop meshcentral as u did, then start meshcentral in the debug le mode `node node_modules/meshcentral --debug cert` and let's see what the output is now

**Reply by Skyline643** [reply]
I've uncommented the letsencrypt section in the config file, with the production value set as 'false'.

root@myVM:~# node /home/admin/node_modules/meshcentral --debug cert
MeshCentral HTTP redirection server running on port 80.
CERT: LE: Getting certs from local store (Staging)
CERT: LE: Reading certificate files
CERT: LE: Setting LE cert for default domain.
MeshCentral v1.1.27, WAN mode.
MeshCentral Intel(R) AMT server running on mydomain.freeddns.org:4433.
MeshCentral HTTPS server running on mydomain.freeddns.org:443, alias port 4433.
CERT: LE: Certificate has 89 day(s) left.
CERT: LE: Certificate is ok.

![image](https://github.com/user-attachments/assets/81b0a769-b98a-4295-9cb5-fb6c58060f23)

I see now 'staging'. But when I change the production value to 'true', my certificate stays 'invalid' :-(

**Reply by si458** [reply]
Ok so what's the output of the debug command when u set production: true ?

Remember to change stop, start debug

**Reply by Skyline643** [reply]
Ok, now I got some more information after retrying:

root@myVM:/home/admin# node /home/admin/node_modules/meshcentral --debug cert
MeshCentral HTTP redirection server running on port 80.
CERT: LE: Getting certs from local store (Production)
CERT: LE: No certificate files found
MeshCentral v1.1.27, WAN mode.
MeshCentral Intel(R) AMT server running on mydomain.freeddns.org:4433.
MeshCentral HTTPS server running on mydomain.freeddns.org:443, alias port 4433.
CERT: LE: Got no certificates, asking for one now.
CERT: LE: Generating private key...
CERT: LE: Setting up ACME client...
CERT: LE: Creating certificate request...
CERT: LE: Requesting certificate from Let's Encrypt...
CERT: LE: Failed to obtain certificate: Error creating new order :: too many certificates (5) already issued for this exact set of domains in the last 168 hours: mydomain.freeddns.org, retry after 2024-08-10T20:30:35Z: see https://letsencrypt.org/docs/duplicate-certificate-limit/


So I need to wait untill the 11th of august, after the weekend.
I found a way on Dynu to disable IPv6 as well. That might maybe solve the problem.
On Letsdebug I have no errors anymore on the AAAA record because IPv6 is disabled.

**Reply by si458** [reply]
Yes that seems to be the case!

NOW, if u took a backup before, copy the production.key and production.crt from inside the letsencrypt backup, and paste it into the new folder, then again restart meshcentral and let's see what happens

**Reply by Skyline643** [reply]
Ok, so I left the virtual machine (with MeshCentral) about 3 days out of service to have no issues with Letsencrypt.

I changed the config file again after booting, staging was ok, changed the config file again and received production certs in the folder.
Certificate seems to be ok, but is not valid again.


LEEVENTS gives me the following output:
```
> LEEVENTS
8/12/2024 2:16:31 PM - Getting certs from local store (Production)
8/12/2024 2:16:31 PM - Reading certificate files
8/12/2024 2:16:31 PM - Setting LE cert for default domain.
8/12/2024 2:16:36 PM - Certificate has 89 day(s) left.
8/12/2024 2:16:36 PM - Certificate is ok.
8/12/2024 2:17:38 PM - Certificate has 89 day(s) left.
8/12/2024 2:17:38 PM - Certificate is ok.
8/12/2024 2:17:50 PM - Certificate has 89 day(s) left.
8/12/2024 2:17:50 PM - Certificate is ok.
8/12/2024 2:17:50 PM - Certificate has 89 day(s) left.
8/12/2024 2:17:50 PM - Certificate is ok.
8/12/2024 2:19:02 PM - Certificate has 89 day(s) left.
8/12/2024 2:19:02 PM - Certificate is ok.
8/12/2024 2:19:30 PM - Failed to respond to challenge, token: letsdebug-test, table: {}.
8/12/2024 2:19:31 PM - Failed to respond to challenge, token: 6_ogl2JgQT_rBBWqneLOk10ZJbHC3zFQ3zirIN38awo, table: {}.
```
When I browse to the IP or dynDNS address of the server (outside my organisation, with mobile phone) I'm redirected from 80 to 4433 (where the agents log on). I get a message to choose a cert on my phone (2 options), but none of both seems to be accepted.

**Reply by si458** [reply]
you shouldnt be asked to pick a certificate on your mobile device, something isnt right with your mobile device or setup
for starters, can u pick a different aliasPort?
4433 shouldnt be used as thats the AMT port and that requires certificates etc so you might be hitting this port by mistake!

also whats your config.json look like now?

if you visit `http://mymeshdomain.com` it should get redirected to `https://mymeshdomain.com:4443`
(port be different if you change it now)
try it in your private window again and see what SSL cert you get

**Reply by Skyline643** [reply]
My config looks now like:

  "settings": {
    "mariadb": {"host": "localhost","user": "myadmin", "password": "something"},
    "cert": "mychosendns.freeddns.org",
    "WANonly": true,
    "_LANonly": true,
    "_sessionKey": "MyReallySecretPassword1",
    "port": 8443,
    "aliasPort": 9443,
    "_redirPort": 80,
    "_redirAliasPort": 80
  },


When I try to browse to http://ipofmyserver:80 I'm redirected to https://ipofmyserver:9443
A simple telnet to the IP or DNS name on port 80 opens the terminal, so port 80 must be open.

**Reply by si458** [reply]
is ur firewall router/nat now port forwarding TCP/80 to TCP/80 meshcentral ip and TCP/9443 to TCP/8443 meshcentral ip ?

and what is the SSL you now get when you private browse to your dns name?

**Reply by Skyline643** [reply]
In the router I did a forward from 80 to 80 (both TCP and UDP) and also from TCP 9443 to TCP 8443. I can access MeshCentral locally on 8443 but not remote with the dynDNS address on 9443.

If i do a simple telnet, only port 80 seems to be open. It seems like the server with MeshCentral is not opening some ports. The firewall on the server is completely disabled. I don't get it anymore...

**Comment by Skyline643** [comment]
Ok, after a long time of searching, trying, ... I found the issue. The problem was port 4433 that wasn't reachable for Letsencrypt. I changed the ports and re-entered all port forwarding. Everything is working now.

I only have two questions:
- Only the remote website (reachable from WAN or with dynDNS address has the certificate). When I browse locally, the connection is unsafe? Do I need to add my local IP to the config file as well?
- Is there a way to export a daily list (CLI or GUI) from all devices that are offline?

We use this to monitor/remote access camera NVR's. It should be great if I could combine Mesh with other tools. To do ping tests to camera's, monitor the disk state, device state (online/offline), etc.
All recommendations are welcome :-)

**Reply by si458** [reply]
how are you browsing locally? (what url)
you should be access it using the DNS/port number you use externally
you cant use an internal dns name with letsencrypt as they cant verifiy you own it

as for the offline, wow thats a very unique case! not sure how you access the NVRs,
unless u are installing the meshagent directly onto them haha

you can use the `meshctrl.js` and `ListDevices` to get a list of all your devices then u can sort which are offline/online etc
https://ylianst.github.io/MeshCentral/meshctrl/

**Reply by Skyline643** [reply]
Ok thanks. Most devices are running on Ubuntu, so no problem to reach them.
I tried also to run the agent on an Teltonika RUT241 but that's not allowed (because the router can't read the syntax in the meshagent file I think). I thought it should work because the routers are OpenWRT based, but unfortunately...

Another thing: the agents are connecting to a central server. But should it be possible to use the meshagent as well to set up a monitoring system. Let's say, use them as a kind of proxy to reach the devices and other devices which are running in the same LAn network. This will be a very difficult setup, but I'm looking for something like that.

**Reply by si458** [reply]
@Skyline643 you should look into the `meshcmd` https://ylianst.github.io/MeshCentral/meshcmd/
it allows port forwarding from remote devices and devices on different networks
